internal/pkg/runtime/engine/apptainer/prepare_linux.go

```diff
@@ -1050,6 +1050,12 @@ func (e *EngineOperations) setSessionLayer(img *image.Image) error {
 		return nil
 	}
 
+	if e.EngineConfig.File.EnableOverlay == "try" && os.Getuid() == 0 && os.Geteuid() == 0 && (writableImage && hasSIFOverlay) {
+		sylog.Debugf("Trying overlay because we are the root user")
+		e.EngineConfig.SetSessionLayer(apptainerConfig.OverlayLayer)
+		return nil
+	}
```

Comment:
That's changing the meaning of the `enable overlay = try` option, and I don't like that very much.  That was only intended to apply to enabling overlay for bind mounts, it doesn't have anything to do with other overlay features.  That's what the description in apptainer.conf says.  On the other hand, further down in the function it is disallowing overlay of other kinds if that value is set to `no`.  In any case, you're not checking for `yes` so that's not right.

I also don't think that there's any need to explicitly check `os.Geteuid() == 0` when `os.Getuid() == 0` is already checked.

Also I don't think that `(writableImage && hasSIFOverlay)` is the right combination of conditions because there may be other conditions where we have the problem, and this is probably not the best place to insert the check either ... it seems to me that the better place is down lower in place of the check for the "overlay" entry in /proc/filesystems.   I suggest changing the `if has..` line and the following line to be something like this:
```
hasoverlay, _ := proc.HasFilesystem("overlay");
if hasoverlay {
        sylog.Debugf("Overlay seems supported and allowed by kernel")
} else if os.Getuid() == 0 {
        // it's probably missing because a modprobe equivalent is needed, which is done later
        sylog.Debugf("Overlay module not loaded but proceeding because we are the root user")
        hasoverlay = true
}
if hasoverlay {
```
It could be argued that this also isn't the best because the root user is restricted by the configuration and we often allow bypassing it, but that's no different than the behavior that was there, and the default settings allow it.

Reply:
Thanks for the great comment! I totally agree that the code should be changed at the place where overlay file system is checked. 👍🏽